Summarize and Rewrite context menu items missing icons when context menu images for internal clients is enabled

lilyspiniolas opened the pull request:
#### eac62bacc9ee9f7e10e6b5a936ca5b2e06be6062
<pre>
Summarize and Rewrite context menu items missing icons when context menu images for internal clients is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=294051">https://bugs.webkit.org/show_bug.cgi?id=294051</a>
<a href="https://rdar.apple.com/152181501">rdar://152181501</a>

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Added enum class `SymbolType `to indicate whether an image is
private or not, added struct `SymbolNameWithType` to store a symbol&apos;s
name and symbol type. Updated method `symbolNameWithTypeForAction` to
return a `SymbolNameWithType` and renamed the method accordingly. As
a result of these changes, `[NSImage imageWithPrivateSystemSymbolName:]`
is now used to get the images for the summarize and rewrite items, resolving
the issue.

* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::menuForTelephoneNumber):
(WebKit::symbolNameWithTypeForAction):
(WebKit::addImageToMenuItem):
(WebKit::symbolNameForAction): Deleted.

Canonical link: <a href="https://commits.webkit.org/295854@main">https://commits.webkit.org/295854@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/574d85ace2920f7faab33a5f8c91cb60eb8232b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26077 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111528 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56924 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26743 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80765 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109332 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21182 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95954 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61093 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56364 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/90525 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14087 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114388 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33466 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89840 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33830 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92182 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89543 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22845 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34421 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12234 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29051 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33391 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38803 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33137 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36490 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34735 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/Platform/mac/MenuUtilities.mm:
```diff
@@ -160,10 +160,23 @@ - (BOOL)revealContext:(RVPresentingContext *)context shouldUseDefaultHighlightFo
 
-static NSString *symbolNameForAction(const WebCore::ContextMenuAction action, bool useAlternateImage)
+enum class SymbolType : uint8_t {
+    None,
```
How about using `std::optional` for the return type instead?

aprotyas replied:
Agreed, since you don't care about the symbol name for `SymbolType::None`, just return a `std::optional<SymbolNameWithType>`

aprotyas replied:
and `enum class SymbolType : uint8_t { Public, Private };`

pxlcoder commented on Source/WebKit/Platform/mac/MenuUtilities.mm:
```diff
@@ -160,10 +160,23 @@ - (BOOL)revealContext:(RVPresentingContext *)context shouldUseDefaultHighlightFo
+
+static SymbolNameWithType symbolNameWithTypeForAction(const WebCore::ContextMenuAction action, bool useAlternateImage)
 {
+    auto symbolName = ""_s;
```
Which also lets you get rid of this.

aprotyas commented on Source/WebKit/Platform/mac/MenuUtilities.mm:
```diff
@@ -201,154 +208,168 @@ - (BOOL)revealContext:(RVPresentingContext *)context shouldUseDefaultHighlightFo
+        return { { SymbolType::Public, "translate"_s } };
     case WebCore::ContextMenuItemTagUnderline:
-        return [NSMenuItem _systemImageNameForAction:@selector(underline:)];
+        return { { SymbolType::Public, [NSMenuItem _systemImageNameForAction:@selector(underline:)] } };
```
r+, general comment about all the non-nullopt returns, I think returning a SymbolTypeWithName object may suffice.

```suggestion
        return { SymbolType::Public, [NSMenuItem _systemImageNameForAction:@selector(underline:)] };
```

lilyspiniolas replied:
I tried that originally to make it cleaner but it seems to cause a build failure 😕

lilyspiniolas committed:
Summarize and Rewrite context menu items missing icons when context menu images for internal clients is enabled
https://bugs.webkit.org/show_bug.cgi?id=294051
rdar://152181501

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Added enum class `SymbolType `to indicate whether an image is
private or not, added struct `SymbolNameWithType` to store a symbol's
name and symbol type. Updated method `symbolNameWithTypeForAction` to
return a `SymbolNameWithType` and renamed the method accordingly. As
a result of these changes, `[NSImage imageWithPrivateSystemSymbolName:]`
is now used to get the images for the summarize and rewrite items, resolving
the issue.

* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::menuForTelephoneNumber):
(WebKit::symbolNameWithTypeForAction):
(WebKit::addImageToMenuItem):
(WebKit::symbolNameForAction): Deleted.

Canonical link: https://commits.webkit.org/295854@main
